lightning/src/ln/channelmanager.rs

```diff
@@ -6207,16 +6276,30 @@ where
 		}
 	}
 
-	fn peer_connected(&self, counterparty_node_id: &PublicKey, init_msg: &msgs::Init) -> Result<(), ()> {
+	fn peer_connected(&self, counterparty_node_id: &PublicKey, init_msg: &msgs::Init, inbound: bool) -> Result<(), ()> {
 		if !init_msg.features.supports_static_remote_key() {
 			log_debug!(self.logger, "Peer {} does not support static remote key, disconnecting with no_connection_possible", log_pubkey!(counterparty_node_id));
 			return Err(());
 		}
 
-		log_debug!(self.logger, "Generating channel_reestablish events for {}", log_pubkey!(counterparty_node_id));
-
 		let _persistence_guard = PersistenceNotifierGuard::notify_on_drop(&self.total_consistency_lock, &self.persistence_notifier);
 
+		// If we have too many peers connected which don't have funded channels, disconnect the
+		// peer immediately. If we have a bunch of unfunded channels taking up space in memory for
+		// disconnected peers, we still let new peers connect, but we'll reject new channels from
+		// them.
+		let mut this_peer_has_funded_channels = false;
+		let connected_peers_with_unfunded_channels = self.peers_with_unfunded_channels(
+			|node_id, node, num_unfunded_channels| {
+				if node_id == counterparty_node_id && num_unfunded_channels != node.channel_by_id.len() {
+					this_peer_has_funded_channels = true;
+				}
+				node.is_connected
+			});
+		if inbound && !this_peer_has_funded_channels && connected_peers_with_unfunded_channels >= MAX_NO_CHANNEL_PEERS {
```

Comment:
IIUC, this enables an adversary to block any new connections to us by simply maintaining 50 open connections with different node IDs.